### ISSUE TITLE
I3Tray.AddSegment's automatic If handling is incorrect (Trac #2136)

### ISSUE THREAD

**Issue by icecube-trac** [icecube-trac/tix2#2128]
`I3Tray.AddSegment` applies conditional functionality to segments which do not support it by attaching state to the `I3Tray` object itself, which modifies the behavior of `I3Tray.AddModule` until the segment returns. However, because this state is not implemented using a stack, it malfunctions when adding a segment which itself adds another segment with an `If` condition attached, as the inner segment overwrites the state associated with the outer, leaving nothing after it completes. 

This is illustrated by the following test case, in which `s2print2` should print only when `s2print1` does, but instead it prints every frame. 

```text
#!/usr/bin/env python

from I3Tray import *
from icecube import icetray

printCounts=dict()
def Printy(frame, name):
	print name,"got frame",frame
	global printCounts
	if not name in printCounts:
		printCounts[name]=1
	else:
		printCounts[name]+=1

@icetray.traysegment
def seg1(tray,name):
	tray.AddModule(Printy,name="s1print")

innerCounter=0
@icetray.traysegment
def seg2(tray,name):
	tray.AddModule(Printy,name="s2print1")
	def odds(frame):
		global innerCounter
		process=(innerCounter%2)==1
		innerCounter+=1
		return process
	tray.AddSegment(seg1,If=odds)
	tray.AddModule(Printy,name="s2print2")
	
tray=I3Tray()
tray.AddModule("BottomlessSource")

itemCounter=0
def AddItem(frame):
	global itemCounter
	frame[str(itemCounter)]=icetray.I3Bool(True)
	itemCounter+=1
tray.AddModule(AddItem)
counter=0
def evens(frame):
	global counter
	process=(counter%2)==0
	counter+=1
	return process
tray.AddSegment(seg2,If=evens)
tray.Execute(8)

print printCounts
```

Migrated from https://code.icecube.wisc.edu/ticket/2136
```json
{
    "status": "closed", 
    "changetime": "2019-03-20T20:36:14", 
    "description": "`I3Tray.AddSegment` applies conditional functionality to segments which do not support it by attaching state to the `I3Tray` object itself, which modifies the behavior of `I3Tray.AddModule` until the segment returns. However, because this state is not implemented using a stack, it malfunctions when adding a segment which itself adds another segment with an `If` condition attached, as the inner segment overwrites the state associated with the outer, leaving nothing after it completes. \n\nThis is illustrated by the following test case, in which `s2print2` should print only when `s2print1` does, but instead it prints every frame. \n\n{{{\n#!/usr/bin/env python\n\nfrom I3Tray import *\nfrom icecube import icetray\n\nprintCounts=dict()\ndef Printy(frame, name):\n\tprint name,\"got frame\",frame\n\tglobal printCounts\n\tif not name in printCounts:\n\t\tprintCounts[name]=1\n\telse:\n\t\tprintCounts[name]+=1\n\n@icetray.traysegment\ndef seg1(tray,name):\n\ttray.AddModule(Printy,name=\"s1print\")\n\ninnerCounter=0\n@icetray.traysegment\ndef seg2(tray,name):\n\ttray.AddModule(Printy,name=\"s2print1\")\n\tdef odds(frame):\n\t\tglobal innerCounter\n\t\tprocess=(innerCounter%2)==1\n\t\tinnerCounter+=1\n\t\treturn process\n\ttray.AddSegment(seg1,If=odds)\n\ttray.AddModule(Printy,name=\"s2print2\")\n\t\ntray=I3Tray()\ntray.AddModule(\"BottomlessSource\")\n\nitemCounter=0\ndef AddItem(frame):\n\tglobal itemCounter\n\tframe[str(itemCounter)]=icetray.I3Bool(True)\n\titemCounter+=1\ntray.AddModule(AddItem)\ncounter=0\ndef evens(frame):\n\tglobal counter\n\tprocess=(counter%2)==0\n\tcounter+=1\n\treturn process\ntray.AddSegment(seg2,If=evens)\ntray.Execute(8)\n\nprint printCounts\n}}}", 
    "reporter": "cweaver", 
    "cc": "", 
    "resolution": "fixed", 
    "_ts": "1553114174597624", 
    "component": "combo core", 
    "summary": "I3Tray.AddSegment's automatic If handling is incorrect", 
    "priority": "normal", 
    "keywords": "", 
    "time": "2018-02-13T15:12:24", 
    "milestone": "Vernal Equinox 2019", 
    "owner": "nwhitehorn", 
    "type": "defect"
}
```

**Comment by icecube-trac** [comment]
Trac update at 20180922T09:16:29: 
* olivas changed milestone from "" to "Winter Solstice 2018"
* olivas changed status from "new" to "assigned"
* olivas changed owner from "" to "olivas"

**Comment by icecube-trac** [comment]
Trac update at 20181204T01:15:44: olivas changed status from "assigned" to "accepted"

**Comment by icecube-trac** [comment]
Trac update at 20190311T14:59:40: 
* nwhitehorn changed status from "accepted" to "assigned"
* nwhitehorn changed owner from "olivas" to "nwhitehorn"

**Comment by icecube-trac** [comment]
Trac update at 20190320T20:36:14: 
* nwhitehorn changed resolution from "" to "fixed"
* nwhitehorn changed status from "assigned" to "closed"
* nwhitehorn commented:

> In changeset 172193:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="172193"
> Fix #2128.
> 
> When adding tray segments, I3Tray automatically propagates If into the
> modules added by the segment. However, it used a single global variable
> that could be overridden if that segment added another segment with a
> different condition (or that had a keyword argument named "If"), resulting
> in generalized space madness. Fix this by maintain a depth-indexed record
> of conditions.
> ```
